Resources.md

```diff
@@ -0,0 +1,118 @@
+
+# MAST : Merkelized Abstract Syntax Tree
+
+Although the MAST idea has been merged into Taproot, its important to get MAST context to well understand Taproot design.
+
+MAST is a combination of a compiler construction (an AST) and a cryptographic one (a Merkle Tree) to gain both privacy and expressivity for Script.
+A gentle [introduction](https://bitcointechtalk.com/what-is-a-bitcoin-merklized-abstract-syntax-tree-mast-33fdf2da5e2f) by David Harding or a [more-in-depth introduction](http://diyhpl.us/wiki/transcripts/bitcoin-core-dev-tech/2017-09-07-merkleized-abstract-syntax-trees/) while a core-dev meetup.
+
+[Merkle Trees](https://rubin.io/public/pdfs/858report.pdf) in themselves have been hard to get right as [some vulnerabilities](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/attachments/20190225/a27d8837/attachment-0001.pdf) affecting the one used in Bitcoin consensus shows it.
+
+The MAST debate has been focus on the [template-vs-generic approach](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-January/015623.html) for script upgrades.
+
+## Generic-design : Merkle branch verification & tail-call semantics for generalized MASTs by maaku
+
+Mark Friedenbach proposed a [script-based approach](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-September/014932.html) relying on the combination of modular mechanism to enable powerful MAST:
+* [BIP 98 : Fast Merkle Tree](https://github.com/bitcoin/bips/blob/master/bip-0098.mediawiki) (maybe relevant in the discussion about Taproot merkle tree) and [ml thread](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-September/014935.html)
+* [BIP 116 : MERKLEBRANCHVERIFY](https://github.com/bitcoin/bips/blob/master/bip-0116.mediawiki)
+* [BIP 117 Tail Call Execution Semantics](https://github.com/bitcoin/bips/blob/master/bip-0117.mediawiki) and [ml thread](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-January/015530.html)
+
+## Templated-design : Merkelized Abstract Syntax Tree
+
+Johnson Lau on a templated-design a la P2SH via a Segwit bump version.
+* [BIP 114 Merkelized Abstract Syntax Tree](https://github.com/bitcoin/bips/blob/master/bip-0114.mediawiki) and [ml announcement](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-September/014963.html) (with examples of further upgrades like re-enabling OP_CAT or OPMUL)
+
+Further debtate on both alternatives and taproot interaction during this [core-dev meetup](https://diyhpl.us/wiki/transcripts/bitcoin-core-dev-tech/2018-03-06-merkleized-abstract-syntax-trees-mast/).
+
+## Lightning
+
+MAST-specific advantages for Lightning:
+- in cooperative closing, make output and spending indistinguishable from other users
+- feature-gating of advanced feature (like SIGHASH_NOINPUT) inside MAST scripts
+
+# Taproot : Privacy perserving switchable scripting
```

Comment:
s/perserving/-preserving/

Reply:
Dash hasn't been added but seems fine without it. Also, that sed is wrong, should have been `s/ perserving/-preserving/`, tsk tsk tsk.